amaranth/vendor/lattice_nexus.py

```diff
@@ -0,0 +1,675 @@
+            m = Module()
+            if self.default_clk == "OSC":
+                if not hasattr(self, "osc_div"):
+                    raise ValueError("OSC divider (osc_div) must be an integer between 2 "
+                                     "and 128")
+                if not isinstance(self.osc_div, int) or self.osc_div < 2 or self.osc_div > 128:
+                    raise ValueError("OSC divider (osc_div) must be an integer between 2 "
+                                     "and 128, not {!r}"
+                                     .format(self.osc_div))
+                clk_i = Signal()
+                m.submodules += Instance("OSC", p_HF_CLK_DIV=self.osc_div, o_HFCLKOUT=clk_i)
+            else:
+                clk_i = self.request(self.default_clk).i
+            if self.default_rst is not None:
+                rst_i = self.request(self.default_rst).i
+            else:
+                rst_i = Const(0)
+
+            gsr0 = Signal()
+            gsr1 = Signal()
+            # There is no end-of-startup signal on ECP5, but PUR is released after IOB enable, so
+            # a simple reset synchronizer (with PUR as the asynchronous reset) does the job.
+            # LIFCL UPDATE: GSR() has a parameter (SYNCMODE) to perform synchronization of the reset to the user clock
+            m.submodules += [
+                #Instance("FD1S3AX", p_GSR="DISABLED", i_CK=clk_i, i_D=~rst_i, o_Q=gsr0),
+                #Instance("FD1S3AX", p_GSR="DISABLED", i_CK=clk_i, i_D=gsr0,   o_Q=gsr1),
+                ## Although we already synchronize the reset input to user clock, SGSR has dedicated
+                ## clock routing to the center of the FPGA; use that just in case it turns out to be
+                ## more reliable. (None of this is documented.)
+                #Instance("SGSR", i_CLK=clk_i, i_GSR=gsr1),
+                Instance("GSR", p_SYNCMODE="SYNC", i_CLK=clk_i, i_GSR_N=~rst_i),
+            ]
+            # GSR implicitly connects to every appropriate storage element. As such, the sync
+            # domain is reset-less; domains driven by other clocks would need to have dedicated
+            # reset circuitry or otherwise meet setup/hold constraints on their own.
+            m.domains += ClockDomain("sync", reset_less=True)
+            m.d.comb += ClockSignal("sync").eq(clk_i)
+            return m
+
+    # pg. 59 family datasheet
+    _single_ended_io_types = [
+        "LVCMOS33", "LVTTL33", 
+        "LVCMOS25", 
+        "LVCMOS18", "LVCMOS18H",
+        "LVCMOS15", "LVCMOS15H",  
+        "LVCMOS12", "LVCMOS12H", 
+        "LVCMOS10", "LVCMOS10H", "LVCMOS10R", 
+        "SSTL135_I", "SSTL135_II", "SSTL15_I", "SSTL15_II",
+        "HSUL12", "MIPI D-PHY LP Input (LVCMOS12)"
+    ]
+    _differential_io_types = [
+        "LVDS", "LVDSE", "subLVDS", "subLVDSEH", "SLVS", "MIPI D-PHY", "LVCMOS33D",
+        "LVTTL33D", "LVCMOS25D", "SSTLD_I", "SSTL135D_I", "SSTL15D_I", "SSTL15D_II",
+        "HSUL15D_I", "HSUL12D", 
+    ]
```

Comment:
Have you checked this list against the datasheet?

Reply:
yes

Reply:
I don't think this is correct, I would expect things like `MIPI_DPHY` instead of `MIPI D-PHY` (although you could check if radiant actually supports both I doubt it?)

there's a list in nextpnr: https://github.com/YosysHQ/nextpnr/blob/2f509734dff676b7543a574b64a46be4224a5aa4/nexus/io.cc#L26 but this might not be definitive

Reply:
My bad, followed the datasheet a little too closely (https://www.latticesemi.com/view_document?document_id=52780, pg. 59). I'll fix the lowercase, and move `MIPI D-PHY` to `MIPI_DPHY` as shown in nextpnr. Will leave 'MIPI D-PHY LP Input (LVCMOS12)` as is --more so an fyi (that you can lvcmos12 for soft mipi LS, see note 6 in datasheet) than something people would actually use.